### ISSUE TITLE
Can PyCapsule be pickled?

### ISSUE THREAD

**Issue by mmckerns** [uqfoundation/dill#354]
Easy way to produce a PyCapsule object:
```
Python 3.6.10 (default, Dec 21 2019, 11:39:07) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import dill
>>> import ctypes
>>> pycapsule = ctypes.pythonapi.PyCapsule_New
>>> pycapsule.restype = ctypes.py_object
>>> c = pycapsule()
>>> c   
\ufffd" at 0x1040f2330>d\ufffdtcH\ufffd
>>> type(c)
<class 'PyCapsule'>
>>> dill.dumps(c)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python3.6/site-packages/dill-0.3.2.dev0-py3.6.egg/dill/_dill.py", line 265, in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
  File "/Users/mmckerns/lib/python3.6/site-packages/dill-0.3.2.dev0-py3.6.egg/dill/_dill.py", line 259, in dump
    Pickler(file, protocol, **_kwds).dump(obj)
  File "/Users/mmckerns/lib/python3.6/site-packages/dill-0.3.2.dev0-py3.6.egg/dill/_dill.py", line 445, in dump
    StockPickler.dump(self, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/pickle.py", line 496, in save
    rv = reduce(self.proto)
TypeError: can't pickle PyCapsule objects
```

Interface is here: https://github.com/python/cpython/blob/master/Include/pycapsule.h

**Comment by mrh335** [comment]
I am also running into the PyCapsule pickle issue using dill and would really appreciate some help to add this feature into dill. 

I have used the pytricubic package to regress a large dataset and the results are stored in a tricubic object. I would like to store this object so that I can load it in other programs to easily compute values using the regressed parameters stored within this object.

This is key as the computation can take 10 hours on a 40 core machine to create and it can't be created every time a simulation needs to be run which each simulation is a few seconds in length. Therefore, want to do this heavy lifting once ahead of time and reap the benefits later.

**Comment by jagjeet-singh** [comment]
Any update on this one? I'm also blocked by `TypeError: can't pickle PyCapsule objects` 
Python: `3.6.8`
dill: `0.3.1.1`
numpy: `1.18.1`

**Comment by adzenith** [comment]
I managed to repro two cases, one where things worked and one where they didn't. In my case the error exists because of this line:

```
builtin_mod = builtin_mod or 'site-packages' in obj.__file__
```
https://github.com/uqfoundation/dill/blob/master/dill/_dill.py#L1305

If you have numpy installed and the path doesn't have site-packages in it, then you encounter the PyCapsule error.

The line I use to repro is:
```
import dill, import numpy; dill.dumps(numpy.core._multiarray_umath)
```
Note that dumping `numpy.core._multiarray_umath._ARRAY_API` will fail in any context.

**Comment by adzenith** [comment]
Locally I am working around this issue by hacking `site-packages` into the paths of our pip modules, but I don't feel like that's the greatest fix.

**Comment by mmckerns** [comment]
Should be resolved by #382.  Please reopen if it isn't.

**Comment by mmckerns** [comment]
Whoops, not resolved yet.

**Comment by guyumengyue** [comment]
Any Update on this? I'm facing the same issue.

**Comment by anivegesana** [comment]
The PyCapsule object wraps an arbitrary C pointer as a Python object. Although it writing functions to pickle and unpickle a PyCapsule are trivial, the semantic meaning of the pointer can change between the interpreter that you pickle the PyCapsule in and the interpreter that you unpickle it from, so most likely the C pointer will retain the same numerical address, but will point to a garbage value. Also, the C extension that created the PyCapsule likely keepa track of it and frees memory when the PyCapsule goes out of scope, so these objects are ephemeral and may not retain their original meaning.

@mmckerns dill can pickle files, which are also highly dependent on their runtime environment. Should PyCapsule pickling capability be added to dill?

**Comment by mmckerns** [comment]
> @mmckerns dill can pickle files, which are also highly dependent on their runtime environment. Should PyCapsule pickling capability be added to dill?

Yes

**Comment by anivegesana** [comment]
Do you have any test cases or CPython extensions I can use to test functionality? Also I don't need to write functionality for Python 2 CObjects right? In my opinion, they are too outdated to be used by modern Python programs.

https://docs.python.org/2/c-api/cobject.html

**Comment by mmckerns** [comment]
> I don't need to write functionality for Python 2 CObjects right?

If the PR will go in after the pending release, and subsequent drop of 2.7, then no.

> Do you have any test cases or CPython extensions I can use to test functionality?

Not particularly.  I generally first go with the objects in `dill._objects` under `HAS_CTYPES`, `Numpy*`, etc.  CPython extensions and objects created using ctypes, should be no different... so the `ctypes` library, `numpy.ctypeslib` (and etc), and `multiprocessing.sharedctypes` are of the best places to get objects from (including user-defined ones).
